packages/contractkit/src/identity/types.ts

```diff
@@ -0,0 +1,102 @@
+import { findFirst } from 'fp-ts/lib/Array'
+import {
+  chain,
+  Either,
+  fromOption,
+  map,
+  mapLeft,
+  parseJSON,
+  toError,
+  tryCatch,
+} from 'fp-ts/lib/Either'
+import { pipe } from 'fp-ts/lib/pipeable'
+import * as t from 'io-ts'
+
+export enum ClaimTypes {
+  ATTESTATION_SERVICE_URL = 'ATTESTATION_SERVICE_URL',
+  DNS = 'DNS',
+  NAME = 'NAME',
+  PROFILE_PICTURE = 'PROFILE_PICTURE',
+  TWITTER = 'TWITTER',
+}
+
+const UrlType = t.string
+const SignatureType = t.string
+const TimestampType = t.number
+
+const AttestationServiceURLClaimType = t.type({
+  type: t.literal(ClaimTypes.ATTESTATION_SERVICE_URL),
+  timestamp: TimestampType,
+  url: UrlType,
+})
+
+const DnsClaimType = t.type({
+  type: t.literal(ClaimTypes.DNS),
+  timestamp: TimestampType,
+  domain: t.string,
+})
+
+const NameClaimType = t.type({
+  type: t.literal(ClaimTypes.NAME),
+  timestamp: TimestampType,
+  name: t.string,
+})
+
+export const ClaimType = t.union([AttestationServiceURLClaimType, DnsClaimType, NameClaimType])
+export const SignedClaimType = t.type({
+  payload: ClaimType,
+  signature: SignatureType,
+})
+
+export const IdentityMetadataType = t.type({
+  claims: t.array(SignedClaimType),
+})
+
+export type Claim = t.TypeOf<typeof ClaimType>
+export type SignedClaim = t.TypeOf<typeof SignedClaimType>
+export type AttestationServiceURLClaim = t.TypeOf<typeof AttestationServiceURLClaimType>
+export type NameClaim = t.TypeOf<typeof NameClaimType>
+export type IdentityMetadata = t.TypeOf<typeof IdentityMetadataType>
+
+export class ValidationError extends Error {}
+// tslint:disable-next-line: max-classes-per-file
+export class ClaimNotFoundError extends Error {}
+
+export const findClaim = (type: ClaimTypes) => (data: IdentityMetadata) =>
```

Comment:
I absolutely agree, but after fiddling with it, I was not able to either type or implement it in a way where it would work? Do you know the typescript magic necessary to do so?

Reply:
After some fiddling myself:

```ts
export interface AttestationClaimPayload {
  type: ClaimTypes.ATTESTATION_SERVICE_URL
  timestamp: number
  url: string
}

export interface DNSClaimPayload {
  type: ClaimTypes.DNS
  timestamp: number
  domain: string
}

export interface NameClaimPayload {
  type: ClaimTypes.NAME
  timestamp: number
  name: string
}

type ClaimPayload<K extends ClaimTypes> = K extends typeof ClaimTypes.DNS
  ? DNSClaimPayload
  : K extends typeof ClaimTypes.NAME ? NameClaimPayload : AttestationClaimPayload

const isOfType = <K extends ClaimTypes>(type: K) => (
  data: SignedClaim['payload']
): data is ClaimPayload<K> => data.type === type

export const fClaim = <K extends ClaimTypes>(type: K) => (
  data: IdentityMetadata
): ClaimPayload<K> => {
  const payload: ClaimPayload<K> | undefined = data.claims
    .map((c) => c.payload)
    .find(isOfType(type))
  if (!payload) {
    throw new Error('bleh')
  }
  return payload
}
```

The key parts is:
a) type guard function :D `isOfType`
b) conditional types to derive a type from a `ClaimType`

Reply:
Interesting, I would have thought the compiler could do this for me. At least that usually happens with redux for me, but of course `io-ts`'s types might cause that